### ISSUE TITLE
Expose the link check as library

### ISSUE THREAD

**Issue by drahnr** [lycheeverse/lychee#64]
Hey, awesome tool and as a matter of fact [cargo-spellcheck](http://github.com/drahnr/cargo-spellcheck) has a bit of an overlap (I just wasn't aware of `lychee`'s existence!).

It would be awesome if `lychee` could be used as a library as well so that arbitrary cmark content can be fed into it, this would be particularly interesting for not re-implementing the logic as part of https://github.com/drahnr/cargo-spellcheck/pull/113

Thanks!

**Comment by mre** [comment]
Hey @drahnr,
you're in luck. #46 will make that happen and it lands... \*checking notes\*... it has just landed. 🎉 
Can you give it a shot and report back? We'd be happy to add any missing pieces.

**Comment by mre** [comment]
BTW added cargo-spellcheck to https://github.com/analysis-tools-dev/static-analysis.
It's live on the website now in case you want to give it an upvote: https://analysis-tools.dev/tag/rust

**Comment by mre** [comment]
Closing this as library support is available now. We can create new issues for library feature requests.  
We'd love to get your feedback after trying lychee. You can create a new issue or comment here. 😄

**Comment by drahnr** [comment]
There are two issues that I can see right now:

* cargo manifest does not expose a library
* a release is pending to include those changes (`0.3.1` was cut right before #64 landed)

Or maybe I missed smth. Thanks in advance!

**Comment by mre** [comment]
Will cut a new release 0.4.0 after #74 is merged. Then we can test the CD pipeline as well. 😄

**Comment by mre** [comment]
@drahnr can you give it another shot? 0.4.0 is out on now on crates.io.

**Comment by drahnr** [comment]
Cool thanks!

There are few things I think that could be improved upon:

* Reduce the exposed module structure - imho it's better to have fewer modules when each of them only contains a single or two entities
* Provide true errors (one that impl's `Send + Sync + 'static + std::error:Error`) rather than using `anyhow`. I'd recommend `thiserror` :)
* Currently one example is provided for the simple case, it would be nice if that was shown as the entry one for docs.rs and expanded in order to showcase the useage of `InputContent` with `collect_links`.

Other than that, it's nice and simple to use!

**Comment by drahnr** [comment]
Another issue that arises, is regarding the use of `tokio 0.2`, lychee's futures will not run nicely with `tokio 0.3`: 
```
thread 'main' panicked at 'must be called from the context of Tokio runtime configured with either `basic_scheduler` or `threaded_scheduler`', /home/bernhard/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.24/src/task/spawn.rs:131:10
```

**Comment by drahnr** [comment]
See https://github.com/drahnr/cargo-spellcheck/pull/131 and run `cargo run -- linkcheck README.md` to see the effect :)

**Comment by mre** [comment]
Alright! Thanks for trying and for all the wonderful tips.
* `thiserror` will be the best option it seems.
* will flatten the module structure a bit
* We'll probably create an examples folder as is good practice. Nice idea!
* Any idea on how to provide the bye- and character ranges? Would be happy for some sample code and output.

We'd love to move to tokio 0.3, but we're currently blocked on #51 😅

**Comment by drahnr** [comment]
>     * We'll probably create an examples folder as is good practice. Nice idea!
> 
>     * Any idea on how to provide the bye- and character ranges? Would be happy for some sample code and output.

with `cmark` it's pretty much free lunch - this is how you get byte ranges https://github.com/drahnr/cargo-spellcheck/blob/master/src/documentation/markdown.rs#L90-L92 and as `cmark` operates on characters, you essentially need to consume the characters upon the start byte bound and count the chars within the byte range to get a char range.

> We'd love to move to tokio 0.3, but we're currently blocked on #51 sweat_smile

It's not so much using tokio, but making it executor agnostic and not rely on any particular runtime :)

**Comment by mre** [comment]
Thanks for the tip about cmark. I'll look into that.
Making the library executor-independent sounds like a good idea as well. Any instructions for that? If you find the time you could even create a ticket for it. Otherwise I can do it later. 😃

**Comment by mre** [comment]
I went ahead and created #82. @drahnr I think we can close this issue here and move the discussion over to #82. What do you think?
